accepted/targeting-packs-and-runtime-packs.md

```diff
@@ -0,0 +1,286 @@
+
+Furthermore, just like `sdk/*` and `shared/*`, `packs/*` content is either included in the sdk .zip or installed via a native installer with appropriate ref-counting. This is in stark, deliberate contrast to the `sdk/NuGetFallbackFolder` that gets content unzipped from each SDKs .lzma, and that only grows over time no matter what you uninstall. The .lzma will be eliminated entirely and replaced by these packs.
+
+
+# Packs Delivered via NuGet
+
+The global scheme above allows for offline builds in common cases. For example, by bundling .NET Core 3.0 targeting packs with .NET Core 3.0.* SDKs, you can install the 3.0.* SDK and build framework-dependent 3.0 projects without pulling anything from the network. However, if you have projects that require a targeting pack that is not globally installed, then the SDK will instruct NuGet to download during restore and use it from the package cache instead of the global location. For this, a new NuGet package type is introduced that:
+
+* cannot be installed into a project via a standard package reference
+* cannot depend on other packages
+* cannot be depended upon by other packages
+* is not subject to standard nuget resolution rules
+
+Furthermore, when NuGet is used, the assets acquired are not cataloged in project.assets.json as normal assets that are subject to NuGet resolution rules. Insetead, the build will use the same logic as the global/offline case and simply redirect it to a downloaded folder in the NuGet package cache. 
+
+This will be implemented using a new "download only package" feature from NuGet: https://github.com/NuGet/Home/issues/7339
+
+
+## Package names
+
+These packages will not appear in the NuGet gallery, NuGet manager, project.assets.json, runtimeconfig.json, or basically anywhere user-facing. As such, the names are essentially implementation details. The naming requirements are:
+
+1. Must not overlap with traditional framework NuGet package names
+2. Given a shared framework name, one can derive the targeting pack package name
+3. Given a shared framework name + RID, one can derive the runtime pack and apphost pack name
+
+**Proposal**
+ * Targeting pack packages: \<shared framework name\>.Ref
+ * Runtime pack packages: \<shared framework name\>.Runtime.\<rid\>
+ * AppHost pack packages: \<shared framework name\>.Host.\<rid\>
+
+**Examples**
+ * Microsoft.NETCore.App.Ref ("base" targeting pack for base .NET Core)
+ * Microsoft.ASPNetCore.App.Ref (targeting pack for ASP.NET Core)
+ * Microsoft.WindowsDesktop.App.Ref (targeting pack for WPF/WinForms on .NET Core)
+ * NETStandard.Library.Ref (targeting pack for .NET Standard)
+ * Microsoft.NETCore.App.Runtime.linux-x64 ("base" runtime pack for .NET Core for Linux OS and x64 CPU)
+ * Microsoft.NETCore.App.Host.linux-x64 (apphost pack for .NET Core for Linux OS and x64 CPU)
+ * Microsoft.ASPNetCore.App.Runtime.osx-x64 (runtime pack for ASP.NET Core for Mac OS and x64 CPOU)
```

Comment:
I think we should keep using existing package IDs. We already have dozens of packages named `runtime.<rid>.<shared framework name>`, and I don't see any benefit to changing this.

On the internal thread, @nguerrera  said:

> I aimed to pick unique names from the existing packages so that they can have a new package type and be impossible to install.

I don't think we have to pick new package IDs to accomplish this. We make no guarantees that new versions of existing package IDs will work the same way. Simply adding `<packageType>` should make the package impossible to install via the current NuGet Visual Studio UI.

Reply:
I still think we should make a clean break with a totally new package name. We're discussing different gallery experiences, etc. It seems like it only adds confusion and complexity to that if the experience difference is version-dependent.

We should consider unlisting the 3.0-preview packages at some point IMHO so that their history ends with 2.2.x where the traditional nuget packages for platform end.

Reply:
Let's set up another meeting to discuss the options here. The last time we tried to squeeze in some objections to the first quarter of the weekly sync did not go well. I will write up the options and book something.

Reply:
I've been thinking some more this afternoon and I think that maybe we can make it work to use the same package names and even folders and simply change package type. But I want to go over all the consequences to be sure.

1. We will have no safeguard against existing clients trying to install it in the usual way.
* Could we use min client version in nuspec instead to point to one that knows about the new package type?

2. We have to be more careful about how things are presented in the gallery and VS UI. For example, we should show different instructions tab based on version, not provide option to upgrade from old world to new one, etc.

3. We have to be more careful about doing this in steps where we don't break things in between. The POR was to have two sets of packages building in parallel and then make a switch on the consumption side.

Reply:
I've been thinking some more this afternoon and I think that maybe we can make it work to use the same package names and even folders and simply change package type. But I want to go over all the consequences to be sure.

1. We will have no safeguard against existing clients trying to install it in the usual way.

* Could we use min client version in nuspec instead to point to one that knows about the new package type?

1. We have to be more careful about how things are presented in the gallery and VS UI. For example, we should show different instructions tab based on version, not provide option to upgrade from old world to new one, etc.
2. We have to be more careful about doing this in steps where we don't break things in between. The POR was to have two sets of packages building in parallel and then make a switch on the consumption side.

Reply:
These names as proposed in the markdown already are final. We want a clean break so that any experience differentiation for the package does not have to consider versions.